core/network/fabric-interop-cc/contracts/interop/membership.go

```diff
@@ -9,45 +9,334 @@
+	if err != nil {
+		return fmt.Errorf("Local Membership JSON Unmarshal error: %s", err)
+	}
+	callerId := ctx.GetClientIdentity()
+	callerCert, err := callerId.GetX509Certificate()
+	if err != nil {
+		return err
+	}
+	err = checkUserInMembership(callerCert, membershipLocalDecoded)
+	if err != nil {
+		return fmt.Errorf("Caller with identity %+v is not a designated IIN Agent: %+v", callerId, err)
+	}
+
+	counterAttestedMembership, err := decodeCounterAttestedMembership(counterAttestedMembershipSerialized)
+	if err != nil {
+		return fmt.Errorf("Counter Attested Membership Unmarshal error: %s", err)
+	}
+	decodedAttestedMembershipSet, err := base64.StdEncoding.DecodeString(counterAttestedMembership.AttestedMembershipSet)
+	if err != nil {
+		return fmt.Errorf("Attested membership set could not be decoded from base64: %s", err.Error())
+	}
+	var attestedMembershipSet identity.CounterAttestedMembership_AttestedMembershipSet
+	err = protoV2.Unmarshal(decodedAttestedMembershipSet, &attestedMembershipSet)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal attested membership set: %s", err.Error())
+	}
+	decodedForeignMembership, err := base64.StdEncoding.DecodeString(attestedMembershipSet.Membership)
+	if err != nil {
+		return fmt.Errorf("Foreign membership could not be decoded from base64: %s", err.Error())
+	}
+	var foreignMembership common.Membership
+	err = protoV2.Unmarshal(decodedForeignMembership, &foreignMembership)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal membership: %s", err.Error())
+	}
+
+	// Match nonces across all attestations, local and foreign
+	matchedNonce := ""
+	for _, attestation := range append(counterAttestedMembership.Attestations, attestedMembershipSet.Attestations...) {
+		if matchedNonce == "" {
+			matchedNonce = attestation.Nonce
+		} else {
+			if matchedNonce != attestation.Nonce {
+				return fmt.Errorf("Mismatched nonces across two attestations: %s, %s", matchedNonce, attestation.Nonce)
+			}
+		}
+	}
+
+	// Ensure valid attestations from all local IIN Agents
+	for _, attestation := range counterAttestedMembership.Attestations {
+		err = validateAttestation(attestation, counterAttestedMembership.AttestedMembershipSet + attestation.Nonce)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Validate foreign membership cert chains
+	err = validateMemberCertChains(&foreignMembership)
+	if err != nil {
+		return err
+	}
+
+	// Ensure authentic and valid attestations from all foreign IIN Agents
+	for _, attestation := range attestedMembershipSet.Attestations {
+		if attestation.UnitIdentity.SecurityDomain != foreignMembership.SecurityDomain {
+			return fmt.Errorf("Foreign agent security domain %s does not match attested membership security domain %s",
+				attestation.UnitIdentity.SecurityDomain, foreignMembership.SecurityDomain)
+		}
+		foundMemberMatch := false
+		for _, member := range foreignMembership.Members {
+			var isSignerRoot bool
```

Comment:
Isn't the following code same as in functions `checkUserInMembership` or `verifyMemberInSecurityDomain`?

Reply:
Yes. I've refactored the code to avoid repetition.